File: py/aoc2024/day9.py
"""
Day 9: Disk Defragmenter
"""

from itertools import accumulate, zip_longest

SAMPLE_INPUT = """
2333133121414131402
"""


def _rangesum(start: int, size: int) -> int:
    return (2 * start + size - 1) * size // 2


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT)
    1928
    """
    data = list(zip_longest(*((int(c) for c in data if c.isdigit()),) * 2, fillvalue=0))
    end, offset, total = len(data), 0, 0
    for i, (used, free) in enumerate(data):
        if i >= end:
            break
        total += i * _rangesum(offset, used)
        offset += used
        for j in range(end - 1, i, -1):
            used2, free2 = data[j]
            moved = min(free, used2)
            total += j * _rangesum(offset, moved)
            offset += moved
            free -= moved
            if not free:
                data[j] = used2 - moved, free2 + moved
                break
            end = j
        offset += free
    return total


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT)
    2858
    """
    data = list(zip_longest(*((int(c) for c in data if c.isdigit()),) * 2, fillvalue=0))
    offsets = list(accumulate(used + free for used, free in data))
    starts, total = [0 for _ in range(10)], 0
    for i in reversed(range(len(data))):
        offset = offsets[i - 1] if i else 0
        used, _ = data[i]
        for j in range(starts[used], i):
            used2, free = data[j]
            if used <= free:
                offset = offsets[j] - free
                data[j] = used2, free - used
                break
        else:
            j = i
        total += i * _rangesum(offset, used)
        for used in range(used, len(starts)):
            if starts[used] >= j:
                break
            starts[used] = j
    return total


parts = (part1, part2)

File: hs/src/Day9.hs
-- |
-- Module:         Day9
-- Description:    <https://adventofcode.com/2024/day/9 Day 9: Disk Defragmenter>
module Day9 (part1, part2) where

import Control.Monad (ap, foldM, when)
import Control.Monad.ST (runST)
import Data.Char (digitToInt, isDigit)
import Data.List (scanl')
import Data.List.Split (chunksOf)
import Data.Maybe (fromMaybe, listToMaybe)
import Data.Text (Text)
import Data.Text qualified as T (unpack)
import Data.Vector (Vector)
import Data.Vector qualified as V (fromList, thaw)
import Data.Vector.Mutable qualified as MV (length, read, replicate, write)

parse :: Text -> Vector (Int, (Int, Int))
parse input =
  V.fromList $
    flip zip `ap` scanl' (flip $ (+) . uncurry (+)) 0 $
      [ (size, fromMaybe 0 $ listToMaybe free)
      | size : free <- chunksOf 2 $ map digitToInt $ filter isDigit $ T.unpack input
      ]

triRange :: (Integral a) => a -> a -> a
triRange offset size = (2 * offset + size - 1) * size `div` 2

part1 :: Text -> Int
part1 input = runST $ do
  files <- V.thaw $ parse input
  let go i j k
        | i < j = do
            (offset, (used, free)) <- MV.read files i
            go' (offset + used) free i j $! k + i * triRange offset used
        | otherwise = pure k
      go' offset free i j k
        | free > 0 && i + 1 < j = do
            (offset', (used', free')) <- MV.read files $ j - 1
            let moved = min free used'
            if moved < used'
              then do
                MV.write files (j - 1) (offset', (used' - moved, free' + moved))
                go (i + 1) j $! k + (j - 1) * triRange offset moved
              else go' (offset + moved) (free - moved) i (j - 1) $! k + (j - 1) * triRange offset moved
        | otherwise = go (i + 1) j k
  go 0 (MV.length files) 0

part2 :: Text -> Int
part2 input = runST $ do
  files <- V.thaw $ parse input
  starts <- MV.replicate 10 0
  let go k i = do
        (offset, (used, _)) <- MV.read files i
        (j, offset') <- fromMaybe (i, offset) <$> (MV.read starts used >>= go' used i)
        go'' used j
        pure $! k + i * triRange offset' used
      go' used i j
        | j < i = do
            (offset, (used', free)) <- MV.read files j
            if used <= free
              then do
                MV.write files j (offset, (used', free - used))
                Just . (j,) . subtract free . fst <$> MV.read files (j + 1)
              else go' used i $ j + 1
        | otherwise = pure Nothing
      go'' used i
        | used < MV.length starts = do
            j <- MV.read starts used
            when (j < i) $ do
              MV.write starts used i
              go'' (used + 1) i
        | otherwise = pure ()
  foldM go 0 [MV.length files - 1, MV.length files - 2 .. 0]
